### ISSUE TITLE
Need help understanding the steal_yourself configuration for Reality

### ISSUE THREAD

**Issue by lostsoul6** [chika0801/Xray-examples#55]
Hello Dear @chika0801 

I'm trying to create a Reality server based on your config here : https://github.com/chika0801/Xray-examples/tree/main/VLESS-XTLS-uTLS-REALITY/steal_yourself

I have multiple questions regarding the setup and I hope you can answer these .

1 - Does this setup basically steal SSL certificate of a real website like www.lovelive-anime.jp and use that SSL to pass is to REALITY ?

in the nginx.conf file , I have some difficulties :  
```

user nginx;
worker_processes auto;

error_log /var/log/nginx/error.log notice;
pid /var/run/nginx.pid;

events {
    worker_connections 1024;
}

http {
    log_format main '[$time_local] $proxy_protocol_addr "$http_referer" "$http_user_agent"';
    access_log /var/log/nginx/access.log main;

    map $http_upgrade $connection_upgrade {
        default upgrade;
        ""      close;
    }

    map $proxy_protocol_addr $proxy_forwarded_elem {
        ~^[0-9.]+$        "for=$proxy_protocol_addr";
        ~^[0-9A-Fa-f:.]+$ "for=\"[$proxy_protocol_addr]\"";
        default           "for=unknown";
    }

    map $http_forwarded $proxy_add_forwarded {
        "~^(,[ \\t]*)*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*([ \\t]*,([ \\t]*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*)?)*$" "$http_forwarded, $proxy_forwarded_elem";
        default "$proxy_forwarded_elem";
    }

    server {
        listen 80;
        return 301 https://$host$request_uri;
    }

    server {
        listen              127.0.0.1:8003 ssl http2 proxy_protocol;
        set_real_ip_from    127.0.0.1;

        ssl_certificate     /etc/ssl/private/fullchain.cer;
        ssl_certificate_key /etc/ssl/private/private.key;

        ssl_protocols       TLSv1.2 TLSv1.3;
        ssl_ciphers         TLS13_AES_128_GCM_SHA256:TLS13_AES_256_GCM_SHA384:TLS13_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305;

        ssl_session_timeout 1d;
        ssl_session_cache   shared:SSL:10m;
        ssl_session_tickets off;

        ssl_stapling        on;
        ssl_stapling_verify on;
        ssl_trusted_certificate /etc/ssl/private/fullchain.cer;
        resolver            1.1.1.1 valid=60s;
        resolver_timeout    2s;

        location / {
            sub_filter                         $proxy_host $host;
            sub_filter_once                    off;

            set $website                       www.lovelive-anime.jp;
            proxy_pass                         https://$website;
            resolver 1.1.1.1;

            proxy_set_header Host              $proxy_host;

            proxy_http_version                 1.1;
            proxy_cache_bypass                 $http_upgrade;

            proxy_ssl_server_name              on;

            proxy_set_header Upgrade           $http_upgrade;
            proxy_set_header Connection        $connection_upgrade;
            proxy_set_header X-Real-IP         $proxy_protocol_addr;
            proxy_set_header Forwarded         $proxy_add_forwarded;
            proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_set_header X-Forwarded-Host  $host;
            proxy_set_header X-Forwarded-Port  $server_port;

            proxy_connect_timeout              60s;
            proxy_send_timeout                 60s;
            proxy_read_timeout                 60s;
        }
    }
}
```
2 - in the above config I see some parameters like :  

```
ssl_certificate     /etc/ssl/private/fullchain.cer;
ssl_certificate_key /etc/ssl/private/private.key;
ssl_trusted_certificate /etc/ssl/private/fullchain.cer;
```

what are these exactly ? will nginx create them automatically or should I create them ? If I have to create them , how do I do so ?

this is the reality server configuration file : 

```
"inbounds": [
        {
            "listen": "0.0.0.0",
            "port": 443,
            "protocol": "vless",
            "settings": {
                "clients": [
                    {
                        "id": "chika", // 执行 xray uuid 生成，或 1-30 字节的字符串
                        "flow": "xtls-rprx-vision"
                    }
                ],
                "decryption": "none"
            },
            "streamSettings": {
                "network": "tcp",
                "security": "reality",
                "realitySettings": {
                    "show": false, // 若为 true，输出调试信息
                    "dest": "8003",
                    "xver": 1,
                    "serverNames": [ // 客户端可用的 serverName 列表，暂不支持 * 通配符，建议填由 Nginx 加载的 TLS 证书中包含的域名
                        ""
                    ],
                    "privateKey": "2KZ4uouMKgI8nR-LDJNP1_MHisCJOmKGj9jUjZLncVU", // 执行 xray x25519 生成，填 "Private key" 的值
                    "shortIds": [ // 客户端可用的 shortId 列表，可用于区分不同的客户端
                        "6ba85179e30d4fc2" // 0 到 f，长度为 2 的倍数，长度上限为 16，可留空，或执行 openssl rand -hex 8 生成
                    ]
                }
            },
```

3 - what address do I enter for serverNames? do I enter the address I have set here in nginx.conf ?

`set $website                       www.lovelive-anime.jp;`

so like this for example?

"serverNames": [ // 客户端可用的 serverName 列表，暂不支持 * 通配符，建议填由 Nginx 加载的 TLS 证书中包含的域名
                        "www.lovelive-anime.jp"

thank you for your help in advance .

**Comment by chika0801** [comment]
2 ssl_certificate /etc/ssl/private/fullchain.cer.
ssl_certificate_key /etc/ssl/private/private.key.
ssl_trusted_certificate /etc/ssl/private/fullchain.cer.

/etc/ssl/private/fullchain.cer; indicates the location of the full SSL certificate chain file in your own SSL certificate.
/etc/ssl/private/private.key; indicates the location of the private key file in your own SSL certificate
ssl_trusted_certificate This parameter is used in conjunction with the following 2 parameters. You can search for this parameter in the NGINX documentation site.
ssl_stapling on. 
ssl_stapling_verify on.

**Comment by chika0801** [comment]
Steal yourself, meaning you install NGINX on your own VPS and load your own applied SSL certificate in the NGINX configuration file. Your own SSL certificate you have to prepare beforehand. (including applying for your own domain name beforehand)

In Xray's configuration file, "dest": "8003", this port 8003 is what NGINX listens to. So the parameters related to loading SSL certificate are configured in the NGINX configuration.

**Comment by chika0801** [comment]
3 The "serverNames" parameter has nothing to do with the domain name that is included in your own SSL certificate that you load in your NGINX configuration. You can set a value of your choice, just make sure that the client-side value is the same as the server-side value.

If you don't understand, you can also use the domain name included in your SSL certificate.

For example, you have a domain name 123.abc.com that points to the IP 1.2.3.4 of your VPS.

You applied for an SSL certificate using this domain name. The domain name included in this SSL certificate is 123.abc.com.

You can leave 123.abc.com out of the "serverNames" parameter and fill in a.c.net, as long as you ensure that the client's "serverNames" is also a.c.net, you will meet the conditions of the REALITY protocol.

My personal habit is to set it to 123.abc.com, which is the same as the domain name included in my certificate

**Comment by lostsoul6** [comment]
Thank you so much . I fully understood the process . Just one question . In the nginx.conf file , there is no mention of the domain name that I apply for . for example there is no 123.abc.com . does it mean we only use its ssl and we don't need a server block in nginx.conf to show content on it ?

**Comment by chika0801** [comment]
The advantage of stealing from yourself, in this usage, is that the TLS fingerprint on the server side, would not be in GO language. (Xray use GO certificate, TLS fingerprint is GO language)

Due to REALITY protocol requirements, the client must start fingerprint emulation (uTLS).

In China, REALITY protocol using foreign sites SNI, may encounter not connected, or by the firewall interference from time to time can not connect. REALITY protocol would have saved you to apply for a domain name and apply for their own SSL certificate event.

You use your own form of stealing your own, you still have to do the thing of buying a domain name and applying for an SSL certificate. But for us to use in China, to rationalize the stability of some.

**Comment by chika0801** [comment]
> Thank you so much . I fully understood the process . Just one question . In the nginx.conf file , there is no mention of the domain name that I apply for . for example there is no 123.abc.com . does it mean we only use its ssl and we don't need a server block in nginx.conf to show content on it ?

Yes, XRAY's Dest 8003 parameter only goes to the nginx configuration to get the SSL certificate information for the 123.abc.com domain.

As I said earlier you can fill in the "serverNames" parameter in the XRAY configuration with any other value you like.

**Comment by chika0801** [comment]
https://github.com/XTLS/Xray-core/issues/1681#issuecomment-1436655742

Here is the author's answer to a question I asked earlier.

**Comment by lostsoul6** [comment]
> > Thank you so much . I fully understood the process . Just one question . In the nginx.conf file , there is no mention of the domain name that I apply for . for example there is no 123.abc.com . does it mean we only use its ssl and we don't need a server block in nginx.conf to show content on it ?
> 
> Yes, XRAY's Dest 8003 parameter only goes to the nginx configuration to get the SSL certificate information for the 123.abc.com domain.
> 
> As I said earlier you can fill in the "serverNames" parameter in the XRAY configuration with any other value you like.

you said that you prefer to use the SSL domain as serverNames , right ? Doesn't it make your domain visible to GFW ?

Generally , if GFW is watching the traffic in this method of stealing from your self , what can they see in their logs ? If they see the serverNames , wouldn't it be better to use a fake one so they don't see your main domain and IP ?

**Comment by chika0801** [comment]
> You said that you prefer to use the SSL domain as serverNames , right ? Doesn't it make your domain visible to GFW ?

I'm used to this myself. In China, the firewall sees my domain name (via SNI). I don't really care about that. Because it's cheap to apply for a domain name. Only a small part of China is implementing SNI whitelist.


> Generally , if GFW is watching the traffic in this method of stealing from your self , what can they see in their logs ? If they see the serverNames , wouldn't it be better to use a fake one so they don't see your main domain and IP ?

I think the firewall sees my local ip, the ip that accesses my vps, and the domain I use (SNI).

I haven't tried it, and when I use a form like steal myself, the SSL certificate is for my own domain name. However, the "serverNames" parameter in the XRAY configuration does not interfere with filling in other domains. I didn't test this situation, so I can't tell you anymore.

**Comment by lostsoul6** [comment]
Dear @chika0801 

I have tested the steal_yourself config and the results are as followed for my Iranian friends reading this : 

DSL Lines : it connects perfectly and I got full speed with it
MTN Irancell : Download Speed of around 12 Mbps and Upload speed around 0.50 Mbps on speedtest.net ( very low upload speed )
IR-MCI ( HamraheAval ) : Didn't connect at all .

Now I have a question from you my friend . This is my current nginx.conf :  

```
user nginx;
worker_processes auto;

error_log /var/log/nginx/error.log notice;
pid /var/run/nginx.pid;

events {
    worker_connections 1024;
}

http {
    log_format main '[$time_local] $proxy_protocol_addr "$http_referer" "$http_user_agent"';
    access_log /var/log/nginx/access.log main;

    map $http_upgrade $connection_upgrade {
        default upgrade;
        ""      close;
    }

    map $proxy_protocol_addr $proxy_forwarded_elem {
        ~^[0-9.]+$        "for=$proxy_protocol_addr";
        ~^[0-9A-Fa-f:.]+$ "for=\"[$proxy_protocol_addr]\"";
        default           "for=unknown";
    }

    map $http_forwarded $proxy_add_forwarded {
        "~^(,[ \\t]*)*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*([ \\t]*,([ \\t]*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*)?)*$" "$http_forwarded, $proxy_forwarded_elem";
        default "$proxy_forwarded_elem";
    }

    server {
        listen 80;
        return 301 https://$host$request_uri;
    }

    server {
        listen              127.0.0.1:8003 ssl http2 proxy_protocol;
        set_real_ip_from    127.0.0.1;

        ssl_certificate     /etc/nginx/cert.crt;
        ssl_certificate_key /etc/nginx/private.key;

        ssl_protocols       TLSv1.2 TLSv1.3;
        ssl_ciphers         TLS13_AES_128_GCM_SHA256:TLS13_AES_256_GCM_SHA384:TLS13_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305;

        ssl_session_timeout 1d;
        ssl_session_cache   shared:SSL:10m;
        ssl_session_tickets off;

        ssl_stapling        on;
        ssl_stapling_verify on;
        ssl_trusted_certificate /etc/nginx/cert.crt;
        resolver            1.1.1.1 valid=60s;
        resolver_timeout    2s;

        location / {
            sub_filter                         $proxy_host $host;
            sub_filter_once                    off;

            set $website                       www.thespruceeats.com;
            proxy_pass                         https://$website;
            resolver 1.1.1.1;

            proxy_set_header Host              $proxy_host;

            proxy_http_version                 1.1;
            proxy_cache_bypass                 $http_upgrade;

            proxy_ssl_server_name              on;

            proxy_set_header Upgrade           $http_upgrade;
            proxy_set_header Connection        $connection_upgrade;
            proxy_set_header X-Real-IP         $proxy_protocol_addr;
            proxy_set_header Forwarded         $proxy_add_forwarded;
            proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_set_header X-Forwarded-Host  $host;
            proxy_set_header X-Forwarded-Port  $server_port;

            proxy_connect_timeout              60s;
            proxy_send_timeout                 60s;
            proxy_read_timeout                 60s;
        }
    }
}

```

As you notice , I have these : 

```
set $website                       www.thespruceeats.com;
proxy_pass                         https://$website;
```

This part is supposed to redirect requests to webserver to www.thespruceeats.com right ?

I tested the following in my browser but I don't get redirected to anywhere

* http://my-vps-ip :  redirects to https://my-vps-ip
* https://my-vps-ip : shows Chrome error 

This page isn’t working My.VPS.IP.Hidden sent an invalid response.
ERR_INVALID_REDIRECT

* http://the-domain-i-got-ssl-for for example http://123.abc.com redirects to https://www.abc.com and shows the following error on chrome : 

This site can’t be reached www.abc.com unexpectedly closed the connection.
Try:

Checking the connection
[Checking the proxy and the firewall](chrome-error://chromewebdata/#buttons)
[Running Windows Network Diagnostics](javascript:diagnoseErrors())
ERR_CONNECTION_CLOSED

* https://the-domain-i-got-ssl-for for example https://123.abc.com redirects to https://www.abc.com and shows the following error on chrome : 

This site can’t be reached www.abc.com unexpectedly closed the connection.
Try:

Checking the connection
[Checking the proxy and the firewall](chrome-error://chromewebdata/#buttons)
[Running Windows Network Diagnostics](javascript:diagnoseErrors())
ERR_CONNECTION_CLOSED

Basically , there is no scenario that the content of " www.thespruceeats.com " is shown . I think something is wrong here . Am I misundertanding something ? In which scenario the visitor is redirected to www.thespruceeats.com ?

**Comment by chika0801** [comment]
set $website www.thespruceeats.com.
proxy_pass https://$website.

This parameter proxy_pass in NGINX, it means that NGINX reverse proxy of a other website, I think GFW firewall will not view the content of this website. If you are in Iran, I suggest you choose a site that is not from your country. Maybe the choice of this site is not very important.

For example, your domain name is 123.abc.com and you set this domain name to point to your VPS and you use the way you steal yourself.

You type 123.abc.com into your browser, and the browser should open with the content of the www.thespruceeats.com website.

Your next question, does it mean that you can't display the content of the www.thespruceeats.com website by typing 123.abc.com? If so, I'm not sure why. You can try another website address. Because the URL of this reverse proxy is not very important. You can also put a static HTML page disguised as a website yourself, which can only make you feel like your domain 123.abc.com is a normal website. So this little disguise, not important.

**Comment by chika0801** [comment]
DSL Lines : it connects perfectly and I got full speed with it
MTN Irancell : Download Speed of around 12 Mbps and Upload speed around 0.50 Mbps on speedtest.net ( very low upload speed )
IR-MCI ( HamraheAval ) : Didn't connect at all .

DSL Lines ， MTN Irancell ，  IR-MCI

These 3 names, are they the names of your local network operators?

I've heard that some of your carriers are experimenting with, for example, single-threaded access to foreign sites with limited speed.

Some carriers take website IP whitelist.

Your country's firewall is really hellishly difficult, I hope you support the past.

**Comment by lostsoul6** [comment]
@chika0801 

when I open 123.abc.com which is the domain I got the SSL for and is the reality domain , I don't see the content of www.thespruceeats.com and I don't see any redirects as well . I think the nginx.conf has problems . Can you please set my nginx.conf on your server and test it ? would be really thankful if you can figure out why the fake content is not showing . changing to other websites doesn't work either . Could the syntax be wrong ?

Also regarding ISPs in Iran we have : 

1) Home ADSL or VDSL providers like : ICT , Shatel , Parsonline and etc . These internets have less limitations that the next 2 .
2 ) MTN Irancell , which is a Cellular network provider . This is still SNI whitelist based and reality works on it .
3) MCI , which is a Cellular network provider with most users . This is completely limited and you can't get more than 100 Kbps UPLOAD speed on it with any xray config . Nothing works on it . It seems they have implemented IP whitelist .

for the third @RPRX said knows how to bypass it but said if he make it public , they would shut down the whole internet !!! Maybe he talks about IP spoofing ?

**Comment by chika0801** [comment]
> for the third @RPRX said knows how to bypass it but said if he make it public , they would shut down the whole internet !!! Maybe he talks about IP spoofing ?

I also don't know exactly what the RPRX method he's talking about is.

**Comment by chika0801** [comment]
![1](https://github.com/chika0801/Xray-examples/assets/88967758/c24feb6a-e237-4ea4-8e93-b0ffac7a8240)

I used your URL www.thespruceeats.com and got the error in the picture.

It looks like this site cannot be backgenerated successfully by the NGINX configuration we are using now. What is the reason for not being successful, I don't know.

**Comment by lostsoul6** [comment]
@chika0801 

Sorry to bother you again Sir . I have an X-UI server that I have successfully routed Spotify and Netflix through WARP on it . I will send my current configuration below . How do I change the following config so that all traffic is routed through WARP and not only Spotify and Netflix ? Thanks for the help .

```
{
  "log": {
    "loglevel": "warning",
    "access": "./access.log",
    "error": "./error.log"
  },
  "api": {
    "tag": "api",
    "services": ["HandlerService", "LoggerService", "StatsService"]
  },
  "inbounds": [
    {
      "tag": "api",
      "listen": "127.0.0.1",
      "port": 62789,
      "protocol": "dokodemo-door",
      "settings": {
        "address": "127.0.0.1"
      }
    }
  ],
  "outbounds": [
    {
      "protocol": "freedom",
      "settings": {}
    },
    {
      "tag": "blocked",
      "protocol": "blackhole",
      "settings": {}
    },
    {
      "tag": "WARP",
      "protocol": "socks",
      "settings": {
        "servers": [
          {
            "address": "127.0.0.1",
            "port": 40000
          }
        ]
      }
    }
  ],
  "policy": {
    "levels": {
      "0": {
        "statsUserDownlink": true,
        "statsUserUplink": true
      }
    },
    "system": {
      "statsInboundDownlink": true,
      "statsInboundUplink": true
    }
  },
  "routing": {
    "domainStrategy": "IPIfNonMatch",
    "rules": [
      {
        "type": "field",
        "inboundTag": ["api"],
        "outboundTag": "api"
      },
      {
        "type": "field",
        "outboundTag": "blocked",
        "ip": ["geoip:private"]
      },
      {
        "type": "field",
        "outboundTag": "blocked",
        "protocol": ["bittorrent"]
      },
      {
        "type": "field",
        "outboundTag": "blocked",
        "domain": [
          "geosite:category-ads-all",
          "geosite:category-ads",
          "geosite:google-ads",
          "geosite:spotify-ads"
        ]
      },
      {
        "type": "field",
        "outboundTag": "WARP",
        "domain": [
          "geosite:google",
          "geosite:netflix",
          "geosite:spotify",
          "geosite:openai"
        ]
      }
    ]
  },
  "stats": {}
}
```

**Comment by chika0801** [comment]
You are recommended to study this https://xtls.github.io/Xray-docs-next/document/level-1/routing-lv1-part1.html It will introduce you to the workflow of the routing module in the Xray configuration. Currently this is only available in Chinese, you will need to translate it yourself.

> Sorry to bother you again Sir . I have an X-UI server that I have successfully routed Spotify and Netflix through WARP on it . I will send my current configuration below . How do I change the following config so that all traffic is routed through WARP and not only Spotify and Netflix ? Thanks for the help .

If you can modify the configuration file completely manually, the easiest way to do this is to take the WARP outbound module and put it at the top of the "outbounds" as the 1st section.


```
    "outbounds": [
      {
        "tag": "WARP",
        "protocol": "socks",
        "settings": {
          "servers": [
            {
              "address": "127.0.0.1",
              "port": 40000
            }
          ]
        }
      },
      {
        "protocol": "freedom",
        "settings": {}
      },
      {
        "tag": "blocked",
        "protocol": "blackhole",
        "settings": {}
      }
    ],
```

**Comment by chika0801** [comment]
Another way, if you can customize the rules of the routing module, you can write a rule for the purpose of a global ending. For example, 0-65535 requests for all ports, sent to a TAG named WARP.

Because inside the routing module, the matching order is also done from top to bottom.

```
        {
            "type": "field",
            "port": "0-65535",
            "outboundTag": "WARP"
        },
```

```
    "routing": {
      "domainStrategy": "IPIfNonMatch",
      "rules": [
        {
          "type": "field",
          "inboundTag": ["api"],
          "outboundTag": "api"
        },
        {
          "type": "field",
          "outboundTag": "blocked",
          "ip": ["geoip:private"]
        },
        {
          "type": "field",
          "outboundTag": "blocked",
          "protocol": ["bittorrent"]
        },
        {
          "type": "field",
          "outboundTag": "blocked",
          "domain": [
            "geosite:category-ads-all",
            "geosite:category-ads",
            "geosite:google-ads",
            "geosite:spotify-ads"
          ]
        },
        {
            "type": "field",
            "port": "0-65535",
            "outboundTag": "WARP"
        },
        {
          "type": "field",
          "outboundTag": "WARP",
          "domain": [
            "geosite:google",
            "geosite:netflix",
            "geosite:spotify",
            "geosite:openai"
          ]
        }
      ]
    },
```

**Comment by chika0801** [comment]
It is my custom to add WIREGUARD related configuration to the Xray configuration file if I want to use the example WARP.

This is because if you use the WARP CLI directly, you may have problems with the VPS using too much memory.

Of course, there are also users who have responded to the WIREGUARD problem with Xray.

You can set a time each day to restart the Xray program once.

The last method is to use a global takeover of traffic into WARP form, in China usually newcomers use a one-click script.

https://github.com/fscarmen/warp

**Comment by lostsoul6** [comment]
@chika0801 

Thanks for the information . I'm aware that the memory usage will be high and I don't want to use it in production servers and want to try for personal server with 1 user . My issue is the syntax . In the Xray configuration that I sent above , it is defined that the likes of Netflix , Spotify and google sites will use WARP . I want all traffic to use WARP on this server . Can you help me and edit my configuration in a way that I can achieve this ? I tested many syntax but it didn't work .
